### PR TITLE
Fix return type for load, learn in BaseAlgorithm

### LINKED ISSUE #1040: [Bug] Too general a return type for load and learn in BaseAlgorithm
### 🐛 Bug

The return type of methods `.load()` and `.learn()` in `BaseAlgorithm` is annotated as `"BaseAlgorithm"`, which means that for any subclass that does not override the methods with suitable annotations, the type won't check if the user is expecting an instance of the subclass. Example:

```python
def load_ppo(cache_path, venv) -> PPO:
    return PPO.load(cache_path, venv)
```

The code above does not type check ("expected PPO, got BaseAlgorithm").

### To Reproduce

Run type checker on code above.

### Expected behavior

Code should type check.

### Solution

The solution to this issue (happy to submit a PR for it) is to replace the return type with the `Self` type variable introduced in [PEP0673](https://peps.python.org/pep-0673). It is built into python 3.11 but it is also available through the PyPI package `typing_extensions`.

Instead of:
https://github.com/DLR-RM/stable-baselines3/blob/304c17dc78dbaeba77c709ec03c1b3847991018d/stable_baselines3/common/base_class.py#L533-L544

we use
```python
if sys.version_info >= (3, 11):
    from typing import Self
else:
    from typing_extensions import Self

class BaseAlgorithm(ABC):
    ...

    @abstractmethod
    def learn(self, ...) -> Self: 
        ...

    @classmethod
    def load(cls, ...) -> Self:
        ...
```

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by Rocamonde**
## Description
Fixes the return type of `.load()` and `.learn()` methods in `BaseAlgorithm` so that they now use the `Self` type [PEP 0673](https://peps.python.org/pep-0673/) instead of `BaseAlgorithm`, which breaks type checkers for use with any subclass.

## Motivation and Context

 Closes #1040.

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/base_class.py
Comment (araffin) [inline stable_baselines3/common/base_class.py:13]: typing extensions might not be installed (pytype is an optional dependency), I would also keep it optional if possible

Comment (AdamGleave) [reply stable_baselines3/common/base_class.py:13]: Yep, I think to avoid this dependency we have two options.

Simplest but somewhat ugly is just to go with the TypeVar solution described in https://peps.python.org/pep-0673/ that Self is aimed to replace. Perhaps not so bad if we only need to do this in a few places.

Alternatively, can make the import only happen for type evaluation:
```
if typing.TYPE_CHECKING:
    from typing_extensions import Self
```
Then we might just be able to write `"Self"` instead of `Self`? If string literals not supported, you could also do something janky like:
```
if typing.TYPE_CHECKING:
    from typing_extensions import Self
else:
    Self = "BaseAlgorithm"
```

Comment (qgallouedec) [reply stable_baselines3/common/base_class.py:13]: > typing extensions might not be installed

No, typing-extensions is installed, since it is a torch dependency since at least version 1.11.

**Comment by qgallouedec** [comment]
Users may use python >= 3.11. So I suggest that you distinguish between the cases:

```python
if sys.version_info >= (3, 11):
    from typing import Self
else:
    from typing_extensions import Self
```

**Comment by Rocamonde** [comment]
This is what I was doing initially. But if typing extensions is installed anyways, that would work regardless. Though I agree it’s cleaner.

File: stable_baselines3/common/base_class.py
Comment (qgallouedec) [reply stable_baselines3/common/base_class.py:13]: But to keep it clean, I recommend adding it to the required dependencies of SB3.

**Comment by Rocamonde** [comment]
Apparently pytype does not support typing_extensions.Self yet, lol.

**Comment by qgallouedec** [comment]
> Apparently pytype does not support typing_extensions.Self yet, lol.

It's a shame, I opened a question about it: https://github.com/google/pytype/issues/1283

**Comment by AdamGleave** [comment]
Given `Self` is not supported, we should probably go with the TypeVar solution from https://peps.python.org/pep-0673/ ?

**Comment by Rocamonde** [comment]
it seems like it won't be supported for some time. Best to use the TypeVar solution for now, update it later.

**Comment by Rocamonde** [comment]
We might want to open another issue (or not mark the current one as completed) so that we remember to go back to this in a couple of months when support is added.

**Comment by Rocamonde** [comment]
@AdamGleave @qgallouedec what do you think of the current status?

File: setup.py
Comment (araffin) [inline setup.py:85]: this is not needed anymore, no?
and requirement should be conditioned by python version

File: docs/misc/changelog.rst
Comment (AdamGleave) [inline docs/misc/changelog.rst:32]: This needs to be updated now that it uses TypeVar not self.

File: stable_baselines3/common/base_class.py
Comment (AdamGleave) [inline stable_baselines3/common/base_class.py:707]: Was this auto-reformatted? (I don't object to the change just seems unrelated to the rest of the PR).

**Review by AdamGleave — COMMENTED** [review]
Changelog needs to be updated and I agree with araffin can clean up the setup.py. Otherwise looks fine. Happy to merge once that's addressed and I or araffin LGTM.

File: stable_baselines3/common/base_class.py
Comment (Rocamonde) [reply stable_baselines3/common/base_class.py:707]: Yeah, I think vscode formatted it, then the linter didn't roll it back as there's some style flexibility. I actually think it's cleaner this way, so if it looks good to you too I would be happy to leave it.

**Comment by Rocamonde** [comment]
Sounds good. Made those changes. Left the small format change in the `load_from_zip_file()` call as despite it being accidental I think it makes the call more readable now. Let me know your thoughts.

**Comment by qgallouedec** [comment]
You also should remove the unused `import sys`

**Comment by Rocamonde** [comment]
Sorry about that. Just did that now.

**Review by qgallouedec — APPROVED** [review]

File: stable_baselines3/common/base_class.py
Comment (Rocamonde) [reply stable_baselines3/common/base_class.py]: I think you might have reviewed an outdated version; already fixed these comments before :)

**Comment by Rocamonde** [comment]
Was going to merge the approved PR but don't have write access.

**Comment by qgallouedec** [comment]
> Was going to merge the approved PR but don't have write access.

Just wait for the checks to to completed, I'll merge it soon.

**Comment by qgallouedec** [comment]
Should we do the same fix here?

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/common/policies.py#L161

**Comment by Rocamonde** [comment]
Sure, let's do that too.

**Comment by qgallouedec** [comment]
Could you also correct the following lines? I think it is within the scope of this PR.

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/common/off_policy_algorithm.py#L332

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/common/on_policy_algorithm.py#L238

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/a2c/a2c.py#L196

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/ddpg/ddpg.py#L129

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/dqn/dqn.py#L268

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/ppo/ppo.py#L310

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/sac/sac.py#L302

https://github.com/DLR-RM/stable-baselines3/blob/18b29a68e8d5a11d0e98aeea539c247f0a913019/stable_baselines3/td3/td3.py#L218

**Comment by Rocamonde** [comment]
Added those changes.

**Review by qgallouedec — APPROVED** [review]

**Comment by araffin** [comment]
@Rocamonde in case you have time in the coming weeks, could you do a similar PR to our contrib repo?
Otherwise, I will open an issue in that repo not to forget ;)

**Comment by Rocamonde** [comment]
Hey! @araffin are you referring to https://github.com/hill-a/stable-baselines?

**Comment by Rocamonde** [comment]
@araffin I suppose you're referring to the return types of specific algorithms? As I guess the base classes are shared with sb3.

What is a good way to search for all occurrences in the codebase?

**Comment by araffin** [comment]
> I suppose you're referring to the return types of specific algorithms? 

yes, and I mean this repo: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib
It follows the exact same structure as SB3.

> What is a good way to search for all occurrences in the codebase?

i think there will be only one occurence in each algorithm folder (for the learn method).

**Pull request merged**
